hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieHadoopFsRelationFactory.scala

```diff
@@ -424,15 +442,16 @@ abstract class BaseHoodieCopyOnWriteIncrementalHadoopFsRelationFactory(override
 
 }
 
-class HoodieCopyOnWriteIncrementalHadoopFsRelationFactory(override val sqlContext: SQLContext,
-                                                          override val metaClient: HoodieTableMetaClient,
-                                                          override val options: Map[String, String],
-                                                          override val schemaSpec: Option[StructType],
-                                                          isBootstrap: Boolean)
-  extends BaseHoodieCopyOnWriteIncrementalHadoopFsRelationFactory(sqlContext, metaClient, options, schemaSpec, isBootstrap) {
+abstract class HoodieCopyOnWriteIncrementalHadoopFsRelationFactory(override val sqlContext: SQLContext,
+                                                                       override val metaClient: HoodieTableMetaClient,
+                                                                       override val options: Map[String, String],
+                                                                       override val schemaSpec: Option[StructType],
+                                                                       isBootstrap: Boolean,
+                                                                       mergeOnReadIncrementalRelation: MergeOnReadIncrementalRelation)
+  extends HoodieBaseCopyOnWriteIncrementalHadoopFsRelationFactory(sqlContext, metaClient, options, schemaSpec, isBootstrap) {
 
   private val incrementalFileIndex = new HoodieIncrementalFileIndex(
-    sparkSession, metaClient, schemaSpec, options, fileStatusCache, false, true)
+    sparkSession, metaClient, schemaSpec, options, fileStatusCache, false, true, mergeOnReadIncrementalRelation)
```

Comment:
do you know why we are using `mergeOnReadIncrementalRelation` within HoodieIncrementalFileIndex even for COW table.

Reply:
shouldn't this be one of `IncrementalRelationV1` or `IncrementalRelationV2`

Reply:
https://github.com/apache/hudi/pull/9888 is the pr that adds this. It seems like it is there because it needs to use util methods. The cow incremental relations don't even have this method

Reply:
That will be complex and I think out of scope of this pr. I can create a followup ticket https://issues.apache.org/jira/browse/HUDI-9623